### PR TITLE
Do the translation of crowbar-specific STONITH config in the rails app

### PR DESCRIPTION
**Pull request by vuntz**
Doing it in the recipes is an issue: as they're default attributes, it means every time we need to look at these attributes, we need to run the recipes. And we will want to not run these recipes all the time in the future. For instance, https://github.com/crowbar/crowbar-core/pull/747

### CODE REVIEW

**Comment by vuntz** [comment]
I just realized we probably need a revision bump to do the changes in a migration too.

**Comment by vuntz** [comment]
@aspiers for reference, that's the bit that was missing for fixing the compute-ha job with https://github.com/crowbar/crowbar-core/pull/747

File: crowbar_framework/app/models/pacemaker_service.rb [hunk 1]
```diff
@@ -681,6 +681,9 @@ def prepare_stonith_attributes(role_attributes, remote_nodes, member_nodes, remo
     cluster_nodes = member_nodes + remote_nodes
     stonith_attributes = role_attributes["stonith"]
 
+    # still make the original mode available
```
Comment (aspiers) [inline crowbar_framework/app/models/pacemaker_service.rb]: I don't get this commit at all.  Please can you expand the commit message to give some context?

File: chef/data_bags/crowbar/template-pacemaker.json [hunk 2]
```diff
@@ -54,7 +54,7 @@
     "pacemaker": {
       "crowbar-revision": 0,
       "crowbar-applied": false,
-      "schema-revision": 32,
+      "schema-revision": 100,
```
Comment (aspiers) [inline chef/data_bags/crowbar/template-pacemaker.json]: Probably a stupid question but why 100?

File: crowbar_framework/app/models/pacemaker_service.rb [hunk 3]
```diff
@@ -676,20 +677,9 @@ def validate_proposal_after_save proposal
...
-    remotes ||= []
-    if remotes.include?(node.name)
-      "remote-#{node["hostname"]}"
-    else
-      node["hostname"]
-    end
-  end
-
-  def prepare_stonith_attributes(role, remote_nodes, member_nodes, remotes, members)
+  def prepare_stonith_attributes(role_attributes, remote_nodes, member_nodes, remotes, members)
```
Comment (aspiers) [inline crowbar_framework/app/models/pacemaker_service.rb]: OK, but IIUC ideally this method signature refactoring would have been done in a separate commit preceding the one introducing the migration, to keep the commits as readable as possible.

File: chef/data_bags/crowbar/template-pacemaker.json (hunk 2 above)
Comment (vuntz) [inline chef/data_bags/crowbar/template-pacemaker.json]: We agreed to bump to 100 for all barclamps for the 4.0 branch a while ago.

File: chef/cookbooks/crowbar-pacemaker/recipes/stonith.rb [hunk 4]
```diff
@@ -47,125 +47,23 @@ def pacemaker_node_name(n)
 when "sbd"
   include_recipe "crowbar-pacemaker::sbd"
 
-# Need to add the hostlist param for shared
+# Nothing to do, translation done in the rails app
```
Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/recipes/stonith.rb]: done

File: crowbar_framework/app/models/pacemaker_service.rb (hunk 1 above)
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service.rb]: I expanded the message. Hopefully it's clearer?

File: crowbar_framework/app/models/pacemaker_service.rb [hunk 5]
```diff
@@ -668,5 +671,97 @@ def validate_proposal_after_save proposal
...
     super
   end
-end
 
+  private
+
+  def pacemaker_node_name(node, remotes)
+    remotes ||= []
+    if remotes.include?(node.name)
+      "remote-#{node["hostname"]}"
```
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service.rb]: AFAIK, it's not duplicated in the rails app. There's a similar method in the cookbook, though.

File: crowbar_framework/app/models/pacemaker_service.rb [hunk 6]
```diff
@@ -668,5 +671,97 @@ def validate_proposal_after_save proposal
...
+    else
+      node["hostname"]
+    end
+  end
+
+  def prepare_stonith_attributes(role, remote_nodes, member_nodes, remotes, members)
```
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service.rb]: I guess we don't have the same definition of "very long" ;-) We can create helper methods for each mode, but I feel it makes the whole thing less readable. How much do you care about it?

File: crowbar_framework/app/models/pacemaker_service.rb (hunk 3 above)
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service.rb]: Done.

**Comment by vuntz** [comment]
Gating needs https://github.com/SUSE-Cloud/automation/pull/1571 to be green

File: crowbar_framework/app/models/pacemaker_service.rb (hunk 6 above)
Comment (aspiers) [inline crowbar_framework/app/models/pacemaker_service.rb]: What?! 90 lines is *disgustingly* long for a method.  Anything greater than one page (i.e. 25 lines) is already bad and has a significant negative impact on legibility, data scoping, testability, modular design etc. etc.  [This website suggests asking questions when methods grow longer than *10* lines.](https://sourcemaking.com/refactoring/smells/long-method)

IMNSHO long methods and bad naming are the top two cardinal sins of programmers.  At least 80% of code I read could be significantly improved by keeping methods short and naming things well.  It baffles me that so often I see even really experienced programmers not doing this.  And my "O" is Not So "H" because it's not really my opinion - it comes from programmers who are WAY cleverer and more experienced than me, e.g. https://youtu.be/8bZh5LMaSmE?t=34s (*strongly* recommend watching the whole video if you haven't seen it already).

So as you can see, you struck a nerve, and I care about this a lot :-)

In this particular case, I do concede that the method mainly consisting of a `case` structure makes the length less bad that a 90-line method would normally be.  But there would be several immediate benefits to splitting it up nevertheless:

- clarify which variables are needed within the `when` bodies (for example, `role` isn't)
- make unit testing of each case cleaner
- make the list of cases immediately visible on one screen
- discourage programmers from growing a huge method even further (this is a really important one, just look at the monstrosity which is `ServiceObject#apply_role`)
- (long-term) open up the possibility for refactoring according to the Open/Closed Principle using a dispatcher and subclasses for each STONITH type, and in fact moving to this from a huge conditional is almost exactly the same redesign shown in the youtube talk linked above!

File: crowbar_framework/app/models/pacemaker_service.rb (hunk 1 above)
Comment (aspiers) [inline crowbar_framework/app/models/pacemaker_service.rb]: Ah OK, yes thanks, this is much clearer now!  However the name `crowbar_mode` is pretty vague, since everything involved here is Crowbar-related.  How about `ui_mode` or `proposal_mode` or similar?

File: crowbar_framework/app/models/pacemaker_service.rb [hunk 7]
```diff
@@ -667,5 +692,105 @@ def validate_proposal_after_save proposal
...
+
+      hypervisor_ip = stonith_attributes["libvirt"]["hypervisor_ip"]
+      hypervisor_uri = "qemu+tcp://#{hypervisor_ip}/system"
+
+      cluster_nodes.each do |cluster_node|
+        stonith_node_name = pacemaker_node_name(cluster_node, remotes)
+
+        # We need to know the domain to interact with for each cluster member; it
+        # turns out the domain UUID is accessible via ohai
+        domain_id = cluster_node["crowbar_ohai"]["libvirt"]["guest_uuid"]
```
Comment (jsuchome) [inline crowbar_framework/app/models/pacemaker_service.rb]: I've just hit a problem when my migration failed because `cluster_node["crowbar_ohai"]` does not contain `libvirt` key. Not sure if my setup was so broken or if we should adapt this code for such eventuality.

Comment (dirkmueller) [inline crowbar_framework/app/models/pacemaker_service.rb]: That means the ohai plugins weren't loaded. This sounds like  a bug elsewhere

Comment (jsuchome) [inline crowbar_framework/app/models/pacemaker_service.rb]: At which point should they be run? This seems to be easily reproducible during the upgrade, right after admin server is upgraded and when crowbar-init is trying to run migrations for new crowbar
